pkg/forward/forward.go

```diff
@@ -384,45 +386,46 @@ func (isdf *InterStepDataForward) forwardAChunk(ctx context.Context) {
 	// activeWatermarkBuffers records the buffers that the publisher has published
 	// a watermark in this batch processing cycle.
 	// it's used to determine which buffers should receive an idle watermark.
-	var activeWatermarkBuffers = make(map[string]bool)
+	var activeWatermarkBuffers = make(map[string][]bool)
 	// forward the highest watermark to all the edges to avoid idle edge problem
 	// TODO: sort and get the highest value
 	for toVertexName, toVertexBufferOffsets := range writeOffsets {
+		activeWatermarkBuffers[toVertexName] = make([]bool, len(toVertexBufferOffsets))
 		if publisher, ok := isdf.publishWatermark[toVertexName]; ok {
 			for index, offsets := range toVertexBufferOffsets {
 				if isdf.opts.vertexType == dfv1.VertexTypeSource || isdf.opts.vertexType == dfv1.VertexTypeMapUDF ||
 					isdf.opts.vertexType == dfv1.VertexTypeReduceUDF {
 					if len(offsets) > 0 {
 						publisher.PublishWatermark(processorWM, offsets[len(offsets)-1], int32(index))
-						activeWatermarkBuffers[toVertexName] = true
+						activeWatermarkBuffers[toVertexName][index] = true
 						// reset because the toBuffer is no longer idling
 						isdf.idleManager.Reset(toVertexName)
 					}
 					// This (len(offsets) == 0) happens at conditional forwarding, there's no data written to the buffer
 				} else { // For Sink vertex, and it does not care about the offset during watermark publishing
 					publisher.PublishWatermark(processorWM, nil, int32(index))
-					activeWatermarkBuffers[toVertexName] = true
+					activeWatermarkBuffers[toVertexName][index] = true
 					// reset because the toBuffer is no longer idling
 					isdf.idleManager.Reset(toVertexName)
 				}
 			}
 		}
 	}
 	// TODO(multi-partition): handle idle watermark publishing for multi partitioned buffer
-	if len(dataMessages) > 0 && len(activeWatermarkBuffers) < len(isdf.publishWatermark) {
```

Comment:
shouldn't we have a different variant of the same check?

Reply:
> I think we should add E2E tests to cover multi-partition. #763

multi-partition is not supported yet. I have just made the refactoring so that it's easy to support multi-partition. Multi-partition support and tests related to it will be a separate pr.